jaclang_jaseci/plugins/common.py

```diff
@@ -0,0 +1,1174 @@
+            DocAccess,
+        ) and (cur_root_id := jctx.get_root_id()):
+            for i in range(1, -1, -1):
+                if cur_root_id in to_root_access.roots[i]:
+                    to_jd.current_access_level = i
+                    return True
+
+        to_jd.current_access_level = None
+        return False
+
+    def __eq__(self, other: object) -> bool:
+        """Override equal implementation."""
+        if isinstance(other, DocArchitype):
+            return self._jac_doc_ == other._jac_doc_
+        elif isinstance(other, DocAnchor):
+            return self._jac_doc_ == other
+
+        return False
+
+    def __hash__(self) -> int:
+        """Override hash implementation."""
+        return self._jac_doc_.__hash__()
+
+
+class NodeArchitype(_NodeArchitype, DocArchitype["NodeArchitype"]):
+    """Overriden NodeArchitype."""
+
+    _jac_type_ = JType.node
+
+    def __init__(self) -> None:
+        """Create node architype."""
+        self._jac_: NodeAnchor = NodeAnchor(obj=self)
+
+    async def connect(self) -> "NodeArchitype":
+        """Return self."""
+        return self
+
+    class Collection(ArchCollection["NodeArchitype"]):
+        """Default NodeArchitype Collection."""
+
+        __collection__ = "node"
+        __indexes__ = [
+            {"fields": [("_id", ASCENDING), ("name", ASCENDING), ("root", ASCENDING)]}
+        ]
+
+        @classmethod
+        def __document__(cls, doc: Mapping[str, Any]) -> "NodeArchitype":
+            """Return parsed NodeArchitype from document."""
+            return cls.build_node(
+                DocAnchor[NodeArchitype](
+                    type=JType.node,
+                    name=cast(str, doc.get("name")),
+                    id=cast(ObjectId, doc.get("_id")),
+                    root=cast(ObjectId, doc.get("root")),
+                    access=DocAccess.from_json(cast(dict, doc.get("access"))),
+                    connected=True,
+                    hashes={
+                        key: hash(dumps(val))
+                        for key, val in doc.get("context", {}).items()
+                    },
+                ),
+                doc,
+            )
+
+    def connect_edge(self, edge: "EdgeArchitype", rollback: bool = False) -> None:
+        """Update DocAnchor that there's newly added edge."""
+        self._jac_doc_.connect_edge(edge._jac_doc_, rollback)
+
+    def disconnect_edge(self, edge: "EdgeArchitype") -> None:
+        """Update DocAnchor that there's edge that has been removed."""
+        self._jac_doc_.disconnect_edge(edge._jac_doc_)
+
+    def allow_node(
+        self, node: Union[DocAnchor, "NodeArchitype"], write: bool = False
+    ) -> None:
+        """Allow target node to access current Architype."""
+        if isinstance(node, DocAnchor):
+            self._jac_doc_.allow_node(node.id, write)
+        elif isinstance(node, NodeArchitype):
+            self._jac_doc_.allow_node(node._jac_doc_.id, write)
```

Comment:
hmmm i think we need to have a diagram for these concepts and their relationships, docanchor, nodearchitype, docarchitype, nodeanchor etc....